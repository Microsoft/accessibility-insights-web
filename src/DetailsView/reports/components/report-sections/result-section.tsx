// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedSFC } from '../../../../common/react/named-sfc';
import { RuleResult } from '../../../../scanner/iruleresults';
import { NoFailedInstancesCongrats } from './no-failed-instances-congrats';
import { InstanceOutcomeType } from './outcome-summary-bar';
import { ResultSectionTitle } from './result-section-title';
import { RuleDetailsGroup, RuleDetailsGroupDeps } from './rule-details-group';

export type ResultSectionDeps = RuleDetailsGroupDeps;

export type ResultSectionProps = {
    deps: ResultSectionDeps;
    rules: RuleResult[];
    containerClassName: string;
    title: string;
    outcomeType: InstanceOutcomeType;
    showDetails?: boolean;
    showCongratsIfNotInstances?: boolean;
    badgeCount: number;
};

export const ResultSection = NamedSFC<ResultSectionProps>('ResultSection', props => {
    const { rules, containerClassName, title, outcomeType, badgeCount, showCongratsIfNotInstances } = props;

    let content = <RuleDetailsGroup deps={props.deps} rules={rules} showDetails={props.showDetails} outcomeType={outcomeType} />;

    if (rules.length === 0 && showCongratsIfNotInstances) {
        content = <NoFailedInstancesCongrats />;
    }

    return (
        <div className={containerClassName}>
            <ResultSectionTitle title={title} count={badgeCount} outcomeType={outcomeType} />
            {content}
        </div>
    );
});
