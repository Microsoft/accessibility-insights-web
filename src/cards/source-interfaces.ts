// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DictionaryStringTo } from '../types/common-types';

interface ScanResults {
    errorCount: number;
    errors: ScanResult[];
    // omitting outputFile as we might not need that
}

interface ScanResult {
    ruleInformation: RuleInfo;
    elementInformation: ElementInfo;
}

interface RuleInfo {
    id: RuleIDEnum; // todo: think about expanding this better
    description: string;
    howToFix: string;
    propertyID: number; // This property is used to link elements with rule violations to relevant documentation.
    condition: string; // not sure about this one
}

interface ElementInfo {
    elementPropertyBag: DictionaryStringTo<string>; // key is a UI Automation property name and the value is the corresponding UI Automation property value.
    patterns: string[];
}

enum RuleIDEnum {}

/** Axe Android */

interface AndroidScanResult {
    // axeConfigurationInfo: AxeAndroidConf; - not sure about this one if its needed or not
    axeContextInfo: AxeContext;
    axeRuleResults: AxeRuleResults[];
}

interface AxeContext {
    deviceInformation: AxeDevice;
    viewInformation: AxeView;
    screenshot: Blob;
}

interface AxeDevice {
    dpi: number;
    name: string;
    osVersion: string;
    screenHeight: number;
    screenWidth: number;
}

interface AxeView {
    viewId: number;
    className: string;
    contentDescription: string;
    isAccessibilityFocusable: boolean;
    isClickable: boolean;
    isEnabled: boolean;
    isImportantForAccessibility: boolean;
    labeledBy: string;
    text: string;
    boundsInScreen: AxeBoundInfo;
    children: AxeView[];
}

interface AxeBoundInfo {
    bottom: number;
    left: number;
    right: number;
    top: number;
}

interface AxeRuleResults {
    axeViewId: number;
    propertyBags: AxeRuleResultProps;
    ruleId: string;
    ruleSummary: string;
    status: string;
}

enum AndroidPropertyBagNames {
    CLASS_NAME = 'className',
    CONTENT_DESCRIPTION = 'contentDescription',
    DPI = 'Screen Dots Per Inch',
    FRAME = 'boundsInScreen',
    HEIGHT = 'height',
    IMPORTANT = 'isImportantForAccessibility',
    IS_CLICKABLE = 'isActive',
    IS_ENABLED = 'isEnabled',
    LABELED_BY = 'labeledBy',
    SPEAKABLE_TEXT = 'Speakable Text',
    WIDTH = 'width',
    EXCEPTION = 'Exception',
    STACK_TRACE = 'Stack Trace',
    EVENT_STREAM = 'Applicable Event Stream',
    ACCESSIBILITY_EVENT = 'AccessibilityEvent',
    IS_TOUCH_STARTED = 'Touch Interaction Started',
    IS_FOCUS_CHANGE_OK = 'Is Focus Change Acceptable',
    IS_TOUCH_EXPLORATION_GESTURE = 'Touch Exploration Started',
    VISIBLE_TEXT = 'Visible Text',
    COLOR_FOREGROUND = 'Foreground Color',
    COLOR_BACKGROUND = 'Background Color',
    COLOR_CONTRAST = 'Color Contrast Ratio',
    CONFIDENCE = 'Confidence in Color Detection',
    SCREEN_HEIGHT = 'Screen Height',
    SCREEN_WIDTH = 'Screen Width',
}

type AxeRuleResultProps = { [key in AndroidPropertyBagNames]: string };
