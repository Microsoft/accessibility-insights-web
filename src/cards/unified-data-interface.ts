// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// this is similar to `TestEngine` interface from axe-core
interface ScanEngine {
    name: string;
    version: string;
}

interface OSData {
    name: string;
    osVersion?: string;
    userAgent?: string;
}

interface ViewPortData {
    platformHeight?: number;
    platformWidth?: number;
}

interface PlatformData {
    osInfo: OSData;
    viewPortInfo: ViewPortData;
}

interface ToolData {
    scanEngine: ScanEngine;
}

interface ScanMetaData {}

interface ScanResults {
    results: UnifiedResults;
    platformInfo: PlatformData;
    toolInfo: ToolData;
    scanMetaData: ScanMetaData;
}

interface UnifiedResults {
    passes: UnifiedResultInstance[];
    violations: UnifiedResultInstance[];
    incomplete: UnifiedResultInstance[];
}

interface UnifiedRuleResult {
    id: string;
    ruleDescription: string;
}

interface InstancePropertyBag<T> {
    [property: string]: T;
}

type StoredInstancePropertyBag = InstancePropertyBag<any>;

interface UnifiedResultInstance {
    id: string;
    evaluation: StoredInstancePropertyBag;
    identifiers: StoredInstancePropertyBag;
    descriptors: StoredInstancePropertyBag;
    resolution: StoredInstancePropertyBag;
}
