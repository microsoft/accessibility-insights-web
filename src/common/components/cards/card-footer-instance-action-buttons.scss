// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import '../../styles/colors.scss';

.kebab-menu-icon {
    flex-grow: 1; // Center the custom SVG
    padding-left: 9px !important;
    padding-right: 9px !important;
    padding-top: 6px !important;
    padding-bottom: 6px !important;
    @media screen and (forced-colors: active) {
        color: buttontext !important;
    }

    &:hover {
        border-style: solid !important;
        border-width: 1px;
        padding-left: 8px !important;
        padding-right: 8px !important;
        padding-top: 5px !important;
        padding-bottom: 5px !important;
    }
}

.kebab-menu-button {
    margin-right: 8px;

    // Since the menu uses a custom SVG, set the button shape:
    width: 32px;
    height: 32px;

    &:hover {
        background-color: $menu-item-background-hover;
    }

    &:global(.is-expanded) {
        background-color: $menu-item-background-active;
    }

    // Add hover and active state:
    &:global(.is-expanded) .kebab-menu-icon,
    &:hover .kebab-menu-icon {
        @media screen and (forced-colors: active) {
            color: highlight;
        }
    }
}
