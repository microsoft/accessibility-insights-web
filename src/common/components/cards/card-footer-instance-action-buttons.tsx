// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ActionButton, DirectionalHint, IButton } from '@fluentui/react';
import { registerIcons } from '@fluentui/react/lib/Styling';
import {
    CardFooterMenuItem,
    CardFooterMenuItemsBuilder,
    CardFooterMenuItemsDeps,
} from 'common/components/cards/card-footer-menu-items-builder';
import { MoreActionsMenuIcon } from 'common/icons/more-actions-menu-icon';
import { NarrowModeStatus } from 'DetailsView/components/narrow-mode-detector';
import { IssueFilingServiceProvider } from 'issue-filing/issue-filing-service-provider';
import * as React from 'react';

import { IssueFilingDialog } from '../../../DetailsView/components/issue-filing-dialog';
import { IssueFilingService } from '../../../issue-filing/types/issue-filing-service';
import { CreateIssueDetailsTextData } from '../../types/create-issue-details-text-data';
import { IssueFilingNeedsSettingsContentProps } from '../../types/issue-filing-needs-setting-content';
import {
    IssueFilingServiceProperties,
    UserConfigurationStoreData,
} from '../../types/store-data/user-configuration-store';
import { IssueFilingButtonDeps } from '../issue-filing-button';
import { Toast, ToastDeps } from '../toast';
import { CardInteractionSupport } from './card-interaction-support';
import styles from './card-kebab-menu-button.scss';

registerIcons({
    icons: {
        MoreActionsMenuIcon: <MoreActionsMenuIcon />,
    },
});

export type CardFooterInstanceActionButtonsDeps = {
    cardInteractionSupport: CardInteractionSupport;
    cardFooterMenuItemsBuilder: CardFooterMenuItemsBuilder;
    issueFilingServiceProvider: IssueFilingServiceProvider;
} & IssueFilingButtonDeps &
    ToastDeps &
    CardFooterMenuItemsDeps;

export interface CardFooterInstanceActionButtonsState {
    showNeedsSettingsContent: boolean;
}

export interface CardFooterInstanceActionButtonsProps {
    deps: CardFooterInstanceActionButtonsDeps;
    userConfigurationStoreData: UserConfigurationStoreData;
    issueDetailsData: CreateIssueDetailsTextData;
    kebabMenuAriaLabel?: string;
    narrowModeStatus?: NarrowModeStatus;
}

export class CardFooterInstanceActionButtons extends React.Component<
    CardFooterInstanceActionButtonsProps,
    CardFooterInstanceActionButtonsState
> {
    private toastRef: React.RefObject<Toast>;
    private fileIssueButtonRef: IButton | null;
    private kebabButtonRef: IButton | null;
    constructor(props: CardFooterInstanceActionButtonsProps) {
        super(props);
        this.toastRef = React.createRef();
        this.state = {
            showNeedsSettingsContent: false,
        };
    }

    public render(): JSX.Element | null {
        const menuItems = this.getMenuItems();
        if (menuItems.length === 0) {
            return null;
        }

        return (
            // The wrapper has to be a real element, not a <>, because we want the placeholder elements
            // the dialog/toast involve to be considered as part of the button for the purposes of layout
            // calculation in this component's parent.
            <div onKeyDown={event => event.stopPropagation()}>
                {this.renderButtons()}
                {this.renderIssueFilingSettingContent()}
                {this.renderCopyFailureDetailsToast()}
            </div>
        );
    }

    public renderButtons(): JSX.Element {
        if (this.props.narrowModeStatus?.isCardFooterCollapsed) {
            return this.renderKebabButton();
        } else {
            return this.renderExpandedButtons();
        }
    }

    public renderKebabButton(): JSX.Element {
        return (
            <ActionButton
                componentRef={ref => (this.kebabButtonRef = ref)}
                className={styles.kebabMenuButton}
                ariaLabel={this.props.kebabMenuAriaLabel || 'More actions'}
                menuIconProps={{
                    iconName: 'MoreActionsMenuIcon',
                    className: styles.kebabMenuIcon,
                }}
                menuProps={{
                    className: styles.kebabMenu,
                    directionalHint: DirectionalHint.bottomRightEdge,
                    shouldFocusOnMount: true,
                    items: this.getMenuItems(),
                }}
            />
        );
    }

    public renderExpandedButtons(): JSX.Element {
        const menuItems = this.getMenuItems();

        return (
            <>
                {menuItems.map(props => (
                    <span key={props.key}>
                        <ActionButton
                            onClick={props.onClick}
                            text={props.text}
                            iconProps={props.iconProps}
                            className={props.key}
                            componentRef={props.componentRef}
                        />
                    </span>
                ))}
            </>
        );
    }

    public renderCopyFailureDetailsToast(): JSX.Element | null {
        const { cardInteractionSupport } = this.props.deps;

        if (!cardInteractionSupport.supportsCopyFailureDetails) {
            return null;
        }

        return <Toast ref={this.toastRef} deps={this.props.deps} />;
    }

    private getMenuItems(): CardFooterMenuItem[] {
        return this.props.deps.cardFooterMenuItemsBuilder.getCardFooterMenuItems({
            ...this.props,
            closeNeedsSettingsContent: this.closeNeedsSettingsContent,
            openNeedsSettingsContent: this.openNeedsSettingsContent,
            toastRef: this.toastRef,
            fileIssueButtonRef: ref => (this.fileIssueButtonRef = ref),
        });
    }

    public renderIssueFilingSettingContent(): JSX.Element | null {
        const { deps, userConfigurationStoreData, issueDetailsData } = this.props;
        const { issueFilingServiceProvider, cardInteractionSupport } = deps;

        if (!cardInteractionSupport.supportsIssueFiling) {
            return null;
        }

        const selectedIssueFilingService: IssueFilingService = issueFilingServiceProvider.forKey(
            userConfigurationStoreData.bugService,
        );
        const selectedIssueFilingServiceData: IssueFilingServiceProperties =
            selectedIssueFilingService.getSettingsFromStoreData(
                userConfigurationStoreData.bugServicePropertiesMap,
            );
        const needsSettingsContentProps: IssueFilingNeedsSettingsContentProps = {
            deps,
            isOpen: this.state.showNeedsSettingsContent,
            selectedIssueFilingService,
            selectedIssueData: issueDetailsData,
            selectedIssueFilingServiceData,
            onClose: this.closeNeedsSettingsContent,
            issueFilingServicePropertiesMap: userConfigurationStoreData.bugServicePropertiesMap,
            afterClosed: this.focusButtonAfterDialogClosed,
        };

        return <IssueFilingDialog {...needsSettingsContentProps} />;
    }

    private focusButtonAfterDialogClosed = (): void => {
        if (this.props.narrowModeStatus?.isCardFooterCollapsed) {
            this.kebabButtonRef?.focus();
        } else {
            this.fileIssueButtonRef?.focus();
        }
    };

    private closeNeedsSettingsContent = (): void => {
        this.setState({ showNeedsSettingsContent: false });
    };

    private openNeedsSettingsContent = (): void => {
        this.setState({ showNeedsSettingsContent: true });
    };
}
