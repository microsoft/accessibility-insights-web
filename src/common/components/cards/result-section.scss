// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
.result-section {
    padding-bottom: 24px;

    .title-container {
        .collapsible-control::before {
            position: relative;
            bottom: 2px;
        }
    }

    > h2 {
        margin: 0px;
        font-size: 17px;
        line-height: 24px;
    }
}
