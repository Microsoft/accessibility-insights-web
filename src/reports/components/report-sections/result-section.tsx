// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NamedFC } from 'common/react/named-sfc';
import * as React from 'react';

import { ResultSectionTitle, ResultSectionTitleProps } from '../../../DetailsView/components/cards/result-section-title';
import { ResultSectionContent, ResultSectionContentDeps, ResultSectionContentProps } from './result-section-content';

export type ResultSectionDeps = ResultSectionContentDeps;

export type ResultSectionProps = ResultSectionContentProps &
    ResultSectionTitleProps & {
        containerClassName: string;
        deps: ResultSectionDeps;
    };

export const ResultSection = NamedFC<ResultSectionProps>('ResultSection', props => {
    const { containerClassName } = props;

    return (
        <div className={containerClassName}>
            <h2>
                <ResultSectionTitle {...props} />
            </h2>
            <ResultSectionContent {...props} />
        </div>
    );
});
