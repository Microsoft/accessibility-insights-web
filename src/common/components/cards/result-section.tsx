// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { css } from '@uifabric/utilities';
import { NamedFC } from 'common/react/named-fc';
import * as React from 'react';

import { TargetAppData } from '../../../common/types/store-data/unified-data-interface';
import { UserConfigurationStoreData } from '../../types/store-data/user-configuration-store';
import { ResultSectionContent, ResultSectionContentDeps, ResultSectionContentProps } from './result-section-content';
import { ResultSectionTitle, ResultSectionTitleProps } from './result-section-title';
import { resultSection } from './result-section.scss';

export type ResultSectionDeps = ResultSectionContentDeps;

export type ResultSectionProps = ResultSectionContentProps &
    ResultSectionTitleProps & {
        userConfigurationStoreData: UserConfigurationStoreData;
        containerClassName: string;
        deps: ResultSectionDeps;
        targetAppInfo: TargetAppData;
        shouldAlertFailuresCount?: boolean;
    };

export const ResultSection = NamedFC<ResultSectionProps>('ResultSection', props => {
    const { containerClassName } = props;

    return (
        <div className={css(containerClassName, resultSection)}>
            <h2>
                <ResultSectionTitle {...props} />
            </h2>
            <ResultSectionContent {...props} />
        </div>
    );
});
