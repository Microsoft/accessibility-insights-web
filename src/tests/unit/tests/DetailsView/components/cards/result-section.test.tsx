// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import * as React from 'react';
import { ResultSection, ResultSectionDeps, ResultSectionProps } from '../../../../../../DetailsView/components/cards/result-section';

describe('ResultSection', () => {
    it('renders', () => {
        const props: ResultSectionProps = {
            containerClassName: 'result-section-class-name',
            deps: {} as ResultSectionDeps,
        } as ResultSectionProps;

        const wrapper = shallow(<ResultSection {...props} />);
        expect(wrapper.getElement()).toMatchSnapshot();
    });
});
