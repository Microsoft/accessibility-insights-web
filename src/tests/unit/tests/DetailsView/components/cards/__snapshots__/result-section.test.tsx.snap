// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultSection renders 1`] = `
<div
  className="result-section-class-name resultSection"
>
  <h2>
    <ResultSectionTitle
      containerClassName="result-section-class-name"
      deps={Object {}}
    />
  </h2>
  <ResultSectionContent
    containerClassName="result-section-class-name"
    deps={Object {}}
  />
</div>
`;
